Basic Lua Scripting Support

NPO-197 opened the pull request:
Here is a proposed starting point to adding Lua script support for MelonDS natively. Currently has bare minimal functionality, as I would like to have others review my work before making to many additions. This is my first project in c++ so please let me know if there are any errors, or parts that I need to re-do. The aim for this is to support accessibility features and trackers, not as a comprehensive debugging tool.

RSDuck commented on src/frontend/qt_sdl/OSD.cpp:
```diff
@@ -438,6 +448,26 @@ void DrawGL(float w, float h)
+    {
+        OverlayCanvas& overlay = *lo;
+        if (overlay.isActive){
+            glGenTextures(1,&overlay.GLtexture);
```
why are you recreating the texture for every draw when you even already have a storage place to keep it across frames?

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,583 @@
+    luaThread = new LuaThread();
+    connect(luaThread,&LuaThread::signalPrint,console,&LuaConsole::onGetText);
+    connect(luaThread,&LuaThread::signalClearConsole,console,&LuaConsole::onClear);
+    connect(buttonPausePlay,&QPushButton::clicked,luaThread,&LuaThread::luaTogglePause);
```
placement of spacing here and in a bunch of other places.

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,583 @@
+}
+
+//Used by "while true framestep" loops
+void LuaThread::luaYeild()
```
is that on purpose or should it be yield?

RSDuck commented on LuaScript.md:
```diff
@@ -0,0 +1,56 @@
+
+`canvas` contains the index of the canavs we just created. `SetCanvas(canvas)` sets the current canavs so all drawing functions know which canvas to draw to.
+
+`function _Update()` is a special function that if defined in a lua script will be called once every frame, we don't need it in this example, but if you need to run some code, or update what is drawn to the screen every frame, this is the way to do it.
```
why are some functions in camelCase and some in PascalCase?

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,583 @@
+int Lua_nextFrame(lua_State* L)
+{
+    frameFlag=false;
+    while(!frameFlag)
```
frameFlag is not volatile, so the compiler will probably skip the loop entirely (and if it doesn't this'll burn a lot of cycles).

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,583 @@
+}
+
+//Is called every frame advance
+void LuaThread::luaUpdate()
```
if I understand everything correctly this will cause a race condition as this function can run simultanously with the emulation of the next frame?

RSDuck reviewed (commented):
The LuaThread is a bit dubious and probably needs some thoughts how this could be implemented better.

NPO-197 commented:
Still some more changes needed but majority of the refactoring is done, no more LuaThread, the Lua state is created and run by the emuThread. This should hopefully address any concerns around potential race conditions. I plan on cleaning up other issues on Sunday.

NPO-197 commented:
I think this should be good for another review, I believe I have addressed all the major issues that have been brought up. If there is anything else that I need to change or fix please let me know. There are still more features / functions I'd like to add, but this is the kind of project that could easily lead to feature creep, so I'm gonna cut myself off here.

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,573 @@
+
+
+extern EmuThread* emuThread;
+LuaConsoleDialog* LuaScript::LuaDialog=nullptr;
```
spacing here and also below

RSDuck commented on src/frontend/qt_sdl/LuaMain.h:
```diff
@@ -0,0 +1,85 @@
+extern lua_State* MainLuaState;
+extern volatile bool FlagPause;
+extern volatile bool FlagStop;
+extern volatile bool FlagNewLua;
```
those don't need to be volatile anymore if everything runs from one thread

RSDuck commented on src/frontend/qt_sdl/OSD.cpp:
```diff
@@ -438,6 +448,36 @@ void DrawGL(float w, float h)
+        }
+        if (!overlay.GLTextureLoaded)
+        {
+            glGenTextures(1,&overlay.GLTexture);
```
this is not what I meant. You only need to create a texture once with glGenTextures and glTexImage2D and then you can repeatedly change the texture's content with glTexSubImage2D.

RSDuck commented on src/frontend/qt_sdl/LuaMain.h:
```diff
@@ -0,0 +1,85 @@
+#include <lua.hpp>
+
+
+class LuaConsole: public QPlainTextEdit
```
I'd move these classes into the namespace

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,573 @@
+//Gets Called once a frame
+void LuaScript::luaUpdate()
+{
+    if(!MainLuaState or FlagPause)
```
using or for || is valid C++?

Peter0x44 replied:
Yes
https://en.cppreference.com/w/cpp/language/operator_alternative
Since C++
You can also use it in C by including iso646.h, but it's just some macros

RSDuck commented on src/frontend/qt_sdl/main.h:
```diff
@@ -272,7 +282,9 @@ private slots:
     void onInsertGBACart();
     void onInsertGBAAddon();
     void onEjectGBACart();
+    void onLuaSaveState(QString filename);
```
use `const QString&` to avoid unnecessary copies.

RSDuck commented on src/frontend/qt_sdl/main.h:
```diff
@@ -272,7 +282,9 @@ private slots:
     void onEjectGBACart();
+    void onLuaSaveState(QString filename);
     void onSaveState();
+    void onLuaLoadState(QString filename);
```
same here

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,573 @@
+}
+
+//Gets Called once a frame
+void LuaScript::luaUpdate()
```
instead of specifying the namespace here you can also just wrap everything in
```
namespace LuaScript
{
...
```

RSDuck commented:
there's probably still some more things, but this is what I found from a glance.

NPO-197 commented:
OK addressed those issues, still some bugs I found that I need to fix, I plan on working on that tomorrow.

NPO-197 commented:
OK ready for another review @RSDuck when you have time.

RSDuck reviewed (approved):
you need to go over all the changes and fix the spacing. Then once the merge conflicts have been solved I guess we can merge it.

NPO-197 commented:
OK went over everything, double checked to make sure the spacing was good. Fixed spelling errors in the markdown. Will probably need someone else to take a look at the merge conflicts at least for the .yml and CMakeLists files.

RSDuck commented:
blah something went wrong while resolving the merge conflict.

RSDuck commented:
To the person who talked about this PR on our IRC:

If you are reading this, we are very sorry, the ban was due to the spam bot for some reason flagging your messages. Let us know how you can communicate again with us.

Arisotura commented:
adding on Generic's message: I removed the ban, so you can come back to the IRC. sorry about it.

NPO-197 commented:
Hey sorry I have been kind of a ghost for a while! It's been over a year since I started this fork, I'm not super up to date on all the changes since then. I'd like to get this project working again with the current version, since I am able to continue working on it now.
I don't have much experience with merging repo's so I will start by asking: 
Is there a way to update this fork so it's not like 100+ commits behind main, or should I try and make a new fork from the current version?
Also I wanted to thank everyone for all the help and feedback so far! I know this isn't exactly a high priority feature, but it's been a wonderful learning experience for me personally, and hopefully others will find it helpful in the future.

RSDuck commented:
there have been a ton of changes on the frontend especially with regards to multi instance within one process support. At this point it would probably smarter to not rebase/merge and instead start fresh while moving over the more independent parts like the Lua interface.

NPO-197 commented:
Understood! I'd prefer that option honestly. Most of the work was on the Lua interpreter side of things, I tried to keep the front end stuff to a minimum while working on this since I figured I'd probably need to remake the front end entirely at some point. So I won't be loosing much.

Also gives me a chance to refresh my memory on how everything works.

I'll make a new fork later today, once that's up I'll add a link to the new one and close this pull request.

jahndan commented:
> adding on Generic's message: I removed the ban, so you can come back to the IRC. sorry about it.

Hello, sorry I just saw this. I don't really know how to use IRC, and I just saw you guys added a link to the discord so I'll probably stick to that since it's more familiar.

I actually started a very rudimentary merge of this branch to main. (It definitely doesn't work, I was just using the merge conflicts to track which parts of the code base I hadn't yet looked at.) Because I've been busy IRL, I haven't touched anything in about 2 weeks, but before that I was mainly trying to get an overview of all the changes made externally to the Lua scripting module itself (i.e. what kind of integration was even necessary to begin with) as well as trying to understand the general structure of the codebase (and also what pieces got refactored, which were many).

> Also gives me a chance to refresh my memory on how everything works.
> I'll make a new fork later today, once that's up I'll add a link to the new one and close this pull request.

I don't see the new fork yet on your profile, but I'll push the half-merge (and some of my notes) I've got so far to mine if you want to take a look (I was in the middle of figuring out how/where to associate a Lua scripting engine for each instance, before starting to refactor the Lua module as a class that could have multiple instances).

ag-advania commented:
Hello, any update on this project ? Thanks !

NPO-197 commented:
Small update for those interested, (thanks to everyone who reached out about this)
Currently I'm waiting on getting a new PC before I continue working on this project, should be able to get started again by the end of September, by then I should have much more free time to continue working on this, and *eventually* get this project to a state where it can be merged with the main project. Thanks again to everyone's support and assistance, as well as everyone's patience.

For those who have reached out asking if they can help in any way, I only ask for a little more time to get a minimal viable version working that is compatible with the recent updates to melonDS. Once that basic frame work is setup, I will be sure to let everyone know, and will then gladly appreciate any help!

gaithern commented:
Hype!

NPO-197 committed:
Next Attempt (2024)!

-Still lots todo! but I have a (mostly) working version up and running!

NPO-197 commented:
Hello! Finally got a new PC and was successful in refactoring this project to work with the current build of melonDS! 
Still a fair bit left before this will be ready to be merged!
## TODO:
- Refactor LuaMain from a namespace to a class, so it can live inside the new EmuInstance instead of just floating off on it's own...
- Fix workflows so it compiles on all platforms
- Test lua scripts on other platforms (besides windows)
- Re-test all the old functions to make sure nothing got broken
- Write up better documentation on how everything works. 

Also thanks to everyone as all-ways for your help! 
If anyone notices anything else that needs to be changed let me know. 😄 
I plan to continue working on this the rest of the week, so there will likely be some more changes incoming.

NPO-197 committed:
New Class "LuaBundle"

Refactored code to make use of new LuaBundle class, this is to bring the implementation closer inline with how EmuThread is implemented, and should now be compatible with multiple emuInstances, if needed. also should make adding support for multiple lua scripts a bit easier as well.

NPO-197 committed:
Update CMakeLists.txt find lua package...

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update vcpkg.json

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
Update flake.nix

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
Update CMakeLists.txt

edo9300 commented:
> if anyone notices anything else that needs to be changed let me know. 😄 I plan to continue working on this the rest of the week, so there will likely be some more changes incoming.

Hey, I noticed this pr, and since I'm pretty experienced with LUA, I decided to give it a look, and at a first glance, I noticed a pretty big issue in your code,
https://github.com/melonDS-emu/melonDS/pull/1671/files#diff-0933c778628624f11db71dc326d6e070849ad62ef2222a917a50590ec90041d1R79 here you're redefining the opaque lua_State pointer used by lua to mean something else, and are later in the code accessing it thus altering the lua_State itself down the line (other than being UB in c/cpp and a odr violation).
```c
c
#define CommonHeader	struct GCObject *next; lu_byte tt; lu_byte marked
struct lua_State {
  CommonHeader;
  lu_byte status;
   [...]
```
This is the begin of the actual definition of `lua_State` (as of lua 5.4.7), if you want to store "global" values, lua offers the `lua_getextraspace` function, that returns an area of memory of a fixed size (by default lua configs it will be `sizeof(void*)`, since melonds targets only 64 bit archs, this will always be 8 bytes) that will be placed right next to the lua_State object itself (so no need to store/retrieve extra info through the lua api) and in case coroutine's are used, this data will be copied to them as well.
When I have more time I could give a more in depth look at it to see if I can find other issues regarding the lua side of things you could've missed.

NPO-197 commented:
@edo9300 
Thanks so much for taking a look at it! 
This is my first big project involving lua, so I super appreciate someone with more experience checking things out! 

I'll look into fixing that right away, and I'll look into the function you suggested. I'm sure you probably just saved me from a big headache in the future!

NPO-197 committed:
Using `lua_getextraspace(` instead of redefining lua_State

Thanks to @edo9300 for catching this!

NPO-197 commented:
OK hopefully that resolves the issue @edo9300 pointed out! 

I may need some help figuring out how to get GitHub actions to compile with the lua library again. 
Other then that, I suppose I can continue working on improving documentation in the meantime, at least until someone points out any other mistakes or potential issues!

NPO-197 committed:
Update CMakeLists.txt & build-ubuntu.yml

Fixed Compile errors on github actions.

NPO-197 commented:
OK fixed CMake to import the lua library properly so it should now compile on Linux / Mac, and got all the GitHub actions working! So the workflows are ready to be approved to double check everything compiles properly. My next goal will be to write up a simple .lua file that can be used as a simple test to make sure everything works as expected on all platforms. 

## TODO:
- [x] ~Refactor LuaMain from a namespace to a class, so it can live inside the new EmuInstance instead of just floating off on it's own...~
- [x] ~Fix workflows so it compiles on all platform~
- [ ] Test lua scripts on other platforms (besides windows)
- [ ] Re-test all the old functions to make sure nothing got broken
- [ ] Write up better documentation on how everything works.

NPO-197 committed:
Fixed some broken functions, added Lua test script to tools

Note I only plan to add files to the tools folder that are helpful for quickly testing the implemented Lua functions. (they aren't used in the build.)

NPO-197 committed:
fixed LuaSaveState

NPO-197 committed:
Update LuaScriptTest.lua

NPO-197 committed:
Fixing qpainter font to work on MacOS

NPO-197 committed:
Update LuaMain.cpp

NPO-197 committed:
Update LuaMain.cpp

NPO-197 committed:
Update LuaMain.cpp

NPO-197 committed:
Update LuaMain.cpp

edo9300 commented:
> Finished testing on Linux and Mac using a VM, everything seems to run as expected on Linux, had to change the default font for one of the drawing functions, since it seemed to not work on Mac. I wrote up a simple Lua script and placed it under tools to test all the different functions, everything seems to be working on all platforms!
> 
> With that finished that's everything on my to-do list crossed off, (sans better documentation) so really all that's needed now is to get someone else to give it a look over and let me know what needs to be improved / changed.

There's another gotcha that i noticed and you should be wary about (and is present in yoru code), lua errors in a c function will corrupt the stack if there are objects with constructors present on the stack frame because lua (unless compiled as c++) uses longjmp/setjmp to handle script errors. This is not an issue on windows because on there longjmp also calls destructors, but is an issue on every other platform. In this pr the only function exhibiting this issue is drawImage https://github.com/melonDS-emu/melonDS/blob/bc40d89fad39862b680ec3bedbcfd65a5db9ff9d/src/frontend/qt_sdl/LuaMain.cpp#L589-L595
If you pass a valid string but invalid other arguments, that QString won't be properly destroyed, so you should move its creation after all the lua_check functions have been executed (since in your case you're not calling anything else that would call into lua and possibly raise lua errors)

NPO-197 committed:
replaced QString with const char* while passing args

NPO-197 commented:
@edo9300 The last push should have resolved the issue you pointed out.

I also took a look at using a C++ compiled version of the lua library, but as you already mentioned on discord since we aren't shipping it as a statically linked library we can't assume that a C++ compiled version of lua will always be available...
 
There may be a way to setup Cmake to compile lua from source ourselves, or something to that effect, but I don't really think that the benefits of doing so would outweigh the added complexity, at least not currently.

NPO-197 committed:
Create Lua_Docs.md

Created a basic Documentation of currently implemented lua functions.

NPO-197 committed:
Update Lua_Docs.md

NPO-197 commented:
Finished writing up the documentation for the currently implemented Lua functions.

RSDuck commented:
Please make the use of whitespaces consistent with the rest of the code.

NPO-197 committed:
Code Clean Up

Round one of clean up

NPO-197 committed:
Merge branch 'master' of https://github.com/NPO-197/melonDS-lua

NPO-197 committed:
Clean LuaMain

Round 1-2 of code clean up

NPO-197 committed:
Code Clean Up

Pass 1-3 should be ready for another round of review.

NPO-197 commented:
> Please make the use of whitespaces consistent with the rest of the code.

Sorry about that, thanks for the reminder @RSDuck!
Did a second pass-through and cleaned up everything to be inline with CONTRIBUTING.md / clean up extraneous whitespaces. 

Let me know if there is anything else I need to go back and take a look at! 😄

NPO-197 committed:
Changed MelonPrint to print,

Renamed "MelonPrint" to just "print",
TIL you can redefine pretty much any lua funct...
Also fixed the lua console not re-opening after closing it the first time.

NPO-197 committed:
Update Lua_Docs.md

NPO-197 committed:
Running Lua Script on the main GUI thread

Switched running the LUA script from the emuThread to the main window / main GUI thread... Should have just done this from the start, this makes things much simpler to implement, and will make future planned GUI functions much easier to add.

NPO-197 committed:
Merge branch 'master' into master

NPO-197 committed:
Code Cleanup

Made for loop more readable in Screen.cpp
Removed include for LuaMain in EmuThread as it is no longer needed.
Cleaned up some whitespace / old commented out code.

NPO-197 committed:
Merge branch 'master' into master

NPO-197 commented:
With that, I should be ready for another review. If @RSDuck or someone has time this weekend to look over it, I'd appreciate it!

RSDuck commented on src/frontend/qt_sdl/EmuInstance.h:
```diff
@@ -279,6 +281,9 @@ class EmuInstance
     bool doAudioSync;
+
+    melonDS::u32 getInputMask(){return inputMask;}
+    std::map<melonDS::u8,bool> KeyboardMask; //For Lua Scripts
```
if this is really always just 256 keys, please use std::bitset or our (better) NonStupidBitfield. Also please avoid always putting 256 into the code.

RSDuck reviewed (approved):
There are still lots of space missing.

We will unfortunately not merge this before the coming 1.0 release, as there are enough new things already which might explode.

RSDuck commented on src/frontend/qt_sdl/Screen.cpp:
```diff
@@ -1083,6 +1136,52 @@ void ScreenPanelGL::osdDeleteItem(OSDItem* item)
+        }
+        glBindTexture(GL_TEXTURE_2D, overlay.GLTexture);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
```
you only need to set the parameters once after the creation of the texture.

RSDuck commented on src/frontend/qt_sdl/LuaMain.cpp:
```diff
@@ -0,0 +1,633 @@
+    luaL_openlibs(L);
+    for (LuaFunction* function : definedLuaFunctions)
+        lua_register(L,function->name,function->cfunction);
+    std::filesystem::current_path(filedir.c_str());
```
I am quite sure this is problematic and leads to issues on Windows with non ASCII characters. Either use wide strings or std::filesystem::u8path.

gaithern commented:
Anyway we can get a compiled version of this branch in the releases page on this repo in the meantime?

AntonioND commented:
Also, could you upload some examples of how to use it?

NPO-197 commented:
@gaithern @AntonioND , made a [test release of the current build](https://github.com/NPO-197/melonDS-lua/releases), you can find the current documentation under the ./tools/LuaScripts/ folder [here.](https://github.com/NPO-197/melonDS-lua/blob/master/tools/LuaScripts/Lua_Docs.md) let me know if you have any trouble!

As of now the currently implemented Lua functions are still rather bare bones, most of the work so far has been just setting up the foundation. If there are specific features / functions that you would find useful to have that aren't implemented yet, then let me know and I can prioritize adding those features next.

I agree with RSDuck about waiting until after the 1.0 release before we think about merging this onto main, It simply doesn't make sense to worry about something that's still very much in the experimental phase like this with such a major release on the horizon.

That being said I do plan to continue working on this in the mean time. 😄 
Thanks as always for everyone's help! 💛

NPO-197 committed:
Merge branch 'melonDS-emu:master' into master

NPO-197 committed:
Revised Keyboard input, currentDir

Changed KeyboardMask to HeldKeys function.
Changed filesystem::current_path to QDir::setCurrent to avoid use of 'c_str'...

NPO-197 committed:
Merge branch 'master' of https://github.com/NPO-197/melonDS-lua

NPO-197 committed:
Merge branch 'melonDS-emu:master' into master

NPO-197 committed:
Merge branch 'master' into master

NPO-197 committed:
Merge branch 'melonDS-emu:master' into master

NPO-197 committed:
Merge branch 'master' into master

NPO-197 committed:
Fixed openGL OSD, added joystick function

NPO-197 committed:
Merge remote-tracking branch 'upstream/master'

NPO-197 committed:
Added ram write, fixed bug with keypress release

NPO-197 committed:
Create Mario64_JoyStick.lua

Added example lua script for a JoyStick in Mario64

NPO-197 committed:
Mouse Wheel support

added support for the mouse wheel in `getMouse`. Added optional spacing argument to `text`.

NPO-197 committed:
Organized LuaFunctions

Organized lua functions into different  libraries,  improved memory read / write functions to be more in line with bizhawk.

NPO-197 committed:
Merge branch 'master' into master

NPO-197 committed:
Added memoryDomain functions

NPO-197 committed:
Merge branch 'master' of https://github.com/NPO-197/melonDS-lua

DarthMDev committed:
Add lua to macOS build dependencies (#5)

NPO-197 committed:
add libfaad-dev dependancy to build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Cmake_adjustment

NPO-197 committed:
cmake_test_2

NPO-197 committed:
cmake_test_3

NPO-197 committed:
cmake_test_4

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
Update build-bsd.yml

NPO-197 committed:
Update build-bsd.yml

NPO-197 committed:
Update LuaSavestate.cpp

Arisotura commented:
I'm gonna check the workflows.

if this is in a mergeable enough state, I guess we can merge it?

and include it in melonDS 1.2

asiekierka commented:
(My PRs fail on macOS too, so I think it might be unrelated.)

Ndymario commented:
I applied the fix from #2527 locally and this builds fine for macOS 👍🏻

NPO-197 committed:
Merge pull request #6 from melonDS-emu/master

macOSFix

NPO-197 committed:
update Documentation

fixed some documentation
added checklist of bizhawk functions currently implemented / planned to implement / not planned to implement

NPO-197 committed:
Merge branch 'master' of https://github.com/NPO-197/melonDS-lua

NPO-197 commented:
O.K. Had a chance to double check things, and updated some of the documentation. I also added a checklist under `tools/LuaScripts` which goes through all the bizhawk Lua functions, and notes which ones are currently implemented, which ones might be useful to implement / change in the future... and which ones that wouldn't be useful to implement or are otherwise outside the scope of this PR.

Seems the macOS check passes now after syncing the recent changes. 

Feel free to go ahead and review everything, the only small change I still want to make would be to include a small warning / disclaimer that Lua scripting support is still early in development.

Also I would like to note currently the biggest difference between my implementation and biz-hawk's Lua scripting, is that in order to run a Lua script continuously in the background, biz-hawk requires that `emu.frameadvance()` is called within an infinite loop. For my implementation the emulator runs the whole script once then checks for an '_Update()" function and calls that function once every frame from there on out. In my opinion it's a less messy way to go about doing it.

It's not too difficult to patch a biz-hawk script so that 'emu.frameadvance()' instead yields from a co-routine, ie one called from '_Update()' *but any scripts currently written for biz-hawk **WILL** need to be changed slightly to work on melonDS*.

I'm not sure it would be worth implementing `emu.frameadvance()` the way it is in bizhawk simply for the sake of compatibility, but I'm open to look into it some more, and I am interested to hear if others would prefer to be able to keep using `emu.frameadvance()` as with biz-hawk.

Other then that I'd say this PR is pretty much ready to be merged! (approval pending...)

goodusername123 commented:
Lua 5.5.0 released recently and it has some nice stuff like lower memory usage, so it might be worth updating the workflow files to point to it? it seems like it's just the ones for BSD and Linux/Ubuntu that need changing although I'm not sure if Lua 5.5.0 has been properly pushed as a package for those systems yet.

NPO-197 commented:
@goodusername123 Thanks for pointing that out, it seems the lua5.5 package is still pending for debian / ubuntu. I don't have any clue how long it will take for the new package to get pushed. My understanding is that since each major revision of lua introduces some non-backwards compatible changes, some package maintainers choose to create a new "lua5.x" package separate from the previous version of lua. Which is different to how the "lua" vcpkg works, which has already been updated to lua5.5.0  (vcpkg is what the mac / windows workflows use) .

So as far as the workflows are concerned, seems we might just have to wait for a lua5.5 package to get added for the Ubuntu builds. TBH messing around with installing packages is something I still struggle to understand, so I'm not sure if there might be other ways to handle lua packages that's more consistent across platforms. :thinking:

NPO-197 committed:
Merge remote-tracking branch 'upstream/master'

NPO-197 committed:
lua 5.5 test

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
Update CMakeLists.txt

NPO-197 committed:
lua5.5 attept 2

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Update build-ubuntu.yml

NPO-197 committed:
Revert "lua 5.5 test"

This reverts commit e5b395e85a5ee7c9172016abc2ec425da2e4b791.

NPO-197 committed:
Fixed lua package name for OpenBSD

Izder456 commented:
> No it's not supposed to do that... 😓 This is what it's supposed to look like, (Here I'm using Metroid Prime, but that shouldn't make a difference...) 
> ...
> What are the video settings you are using?

<img width="530" height="288" alt="image" src="https://github.com/user-attachments/assets/8d9f645a-8cb6-4548-8eec-324041cd0a3f" />

with the logo added to that dir, its a little better.

<img width="2720" height="1024" alt="image" src="https://github.com/user-attachments/assets/5326b8f6-1b1d-4378-8edc-307bec0b9b6b" />

with the classic renderer it does this too.

also- i noticed the program segfaults if you try to load a script before you load a game. we should really have a guard against that.

NPO-197 committed:
Add Safety Check for NDS existing

Prevents segfault when trying to read from NDS memory before any ROM has been loaded. Now we properly throw an error to the Lua console.

NPO-197 committed:
Fixed Overlay shader

Fixes bug with lua overlays not rendering properly when using openGL rendering.

NPO-197 committed:
test script cleaning

 cleaned up lua scripts, as well as documentation. Fixed `DrawImage` and added `DrawImageRegion` to be same as bizhawk. Fixed missing `usememorydomain` function.

NPO-197 commented:
OK the graphical issues turned out to be a simple fix in the end, the openGL shader for the lua overlays just wasn't linking properly due to a small change in how the screens are being rendered that I did not catch until now. I rewrote the test script and cleaned up some of the documentation. 

@Izder456 if you could run another test on OpenBSD to confirm that everything is working properly now that would be a big help. :yellow_heart:

Izder456 commented:
> OK the graphical issues turned out to be a simple fix in the end, the openGL shader for the lua overlays just wasn't linking properly due to a small change in how the screens are being rendered that I did not catch until now. I rewrote the test script and cleaned up some of the documentation.
> 
> @Izder456 if you could run another test on OpenBSD to confirm that everything is working properly now that would be a big help. 💛

Seems to work on my end. Thanks! (Also- you might wanna document the new build depend on lua 5.4 in BUILD.md)

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/2394a280-5a37-42ee-9d97-9c18b5ca6de1" />

NPO-197 committed:
Update BUILD.md

NPO-197 committed:
cleanup

clean up of code to fit contribution guidelines more closely. Fixed some memory functions as well.

NPO-197 committed:
Cmake clean

NPO-197 committed:
Close lua console when main window closes

NPO-197 committed:
Merge pull request #8 from melonDS-emu/master

Sync to main March13

NPO-197 committed:
Merge branch 'master' of https://github.com/NPO-197/melonDS-lua

Veddy1674 commented:
I have [forked this repository](https://github.com/Veddy1674/melonDS-lua) and tried my own way of adding support for Lua scripting, if anyone is still interested, you could look up my alternative, which uses sol2 and lua 5.4, adds a "scripting" button which lets you execute scripts in-game, modules "memory" to read/write memory, "emu" to pause/unpause, save and load states, functions like get_state to get the raw pixels on screen, and as you may realize, it's not focused on TAS-making, as most scripting tools in emulators are, but more on dataset-making for AI training and such, but feel free to look it up and make it fit your needs.

Still experimental though

AntonioND commented:
I've tested this a bit, and I see some issues:

- I need some way to start the script at boot. Something like `melonDS --lua myscript.lua rom.nds`. The ROM should start paused if a Lua script is provided.
- I need some way to control the number of frames that the emulator runs, and keep it paused at any other time.
- I need to at least have some way to create screenshots and to shut down melonDS.

For context, most of my tests involve running the emulator for a few frames, maybe pressing a button and releasing it later, taking a screenshot and closing the emulator.

I'm not sure how feasible all of this is with the current melonDS, though.

I did something like this for my own emulator a few years ago, and I had to modify a few things to get it working: https://github.com/AntonioND/giibiiadvance/blob/ccb40c3cf7d9efec538858a36d4ab69bb4d0ce5c/source/lua_handler.c#L263-L270
